pkg/common/cluster_state.go

```diff
@@ -38,23 +48,25 @@ func (d *DesiredClusterState) AddActions(actions []ClusterAction) DesiredCluster
 }
 
 type ClusterState struct {
-	KeycloakServiceMonitor          *monitoringv1.ServiceMonitor
-	KeycloakPodMonitor              *monitoringv1.PodMonitor
-	KeycloakPrometheusRule          *monitoringv1.PrometheusRule
-	KeycloakGrafanaDashboard        *grafanav1alpha1.GrafanaDashboard
-	DatabaseSecret                  *v1.Secret
-	PostgresqlPersistentVolumeClaim *v1.PersistentVolumeClaim
-	PostgresqlService               *v1.Service
-	PostgresqlDeployment            *v12.Deployment
-	KeycloakService                 *v1.Service
-	KeycloakDiscoveryService        *v1.Service
-	KeycloakDeployment              *v12.StatefulSet
-	KeycloakAdminSecret             *v1.Secret
-	KeycloakIngress                 *v1beta1.Ingress
-	KeycloakRoute                   *v13.Route
-	PostgresqlServiceEndpoints      *v1.Endpoints
-	PodDisruptionBudget             *v1beta12.PodDisruptionBudget
-	KeycloakProbes                  *v1.ConfigMap
+	KeycloakServiceMonitor             *monitoringv1.ServiceMonitor
+	KeycloakPodMonitor                 *monitoringv1.PodMonitor
+	KeycloakPrometheusRule             *monitoringv1.PrometheusRule
+	KeycloakGrafanaDashboard           *grafanav1alpha1.GrafanaDashboard
+	DatabaseSecret                     *v1.Secret
+	PostgresqlPersistentVolumeClaim    *v1.PersistentVolumeClaim
+	PostgresqlService                  *v1.Service
+	PostgresqlDeployment               *v12.Deployment
+	KeycloakService                    *v1.Service
+	KeycloakDiscoveryService           *v1.Service
+	KeycloakDeployment                 *v12.StatefulSet
+	KeycloakAdminSecret                *v1.Secret
+	KeycloakIngress                    *v1beta1.Ingress
+	KeycloakRoute                      *v13.Route
+	PostgresqlServiceEndpoints         *v1.Endpoints
+	PodDisruptionBudget                *v1beta12.PodDisruptionBudget
+	KeycloakProbes                     *v1.ConfigMap
+	MigrateBackup                      *v14.Job
```

Comment:
I believe this should be done on higher level of abstraction - by creating a `KeycloakBackup` CR and waiting till it's in `BackupPhaseCreated` status.

Reply:
Hi @slaskawi , thank you for your comment.

> by creating a KeycloakBackup CR and waiting till it's in BackupPhaseCreated status

In fact, this was also my first choice to implement this ticket.
However, I found it hard to implement. 

As current loop is a Keycloak Reconcile loop, I am not sure how to jump into KeycloakBackup's Reconcile loop without user's trigger. In another word, in which place the new KeycloakBackup CR should be created? How to build the communication between Keycloak_controller and Keycloakbackup_controller?

Reply:
You shouldn't all you need to do is to create a new `KeycloakBackup` CR and wait till https://github.com/keycloak/keycloak-operator/blob/master/pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go#L85 is equal to https://github.com/keycloak/keycloak-operator/blob/master/pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go#L76

No communication between controller is needed. Does this help?

Reply:
Sorry, I think I need more help. I appreciate if you can give me more details.

Because the upgrade is detected by Keycloak_migrator, which is created by keycloak_controller. And Keycloak controller is handling all resources which belong to CR Keycloak. 

However, CR KeycloakBackup belong to another controller: KeycloakBackup_controller. This controller has its Reconciler.

In my opinion, once the keycloak_migrator finds the Keycloak version upgrade happen, we can add some code for keycloak controller to create some resources which belongs to CR Keycloak. Because only Keycloak controller's reconcile loop is working. I am not sure how to create a new CR KeycloakBackup because it is hard to make KeycloakBackup_controller's Reconciler loop work. 

Sorry it is difficult to describe my question clearly. I hope you can understand my question. Thank you.

Reply:
@slaskawi

Reply:
Hi @slaskawi, please kindly check my question, thank you!

Reply:
I see. I think I can work it out now.

Reply:
Hey @JiyeYu I'm very sorry for massive delay. I was side-tracked by other duties and simply didn't manage to get back you. I see you implemented most of the requested changes, so let me review them now.